### PR TITLE
Use openssl 1.1 by default

### PR DESCRIPTION
**Pull request by globin**
Tracking hydra jobset: https://hydra.nixos.org/jobset/nixpkgs/openssl-1.1

### CODE REVIEW

**Comment by vcunat** [comment]
I guess this isn't yet ready to be staged; it seems many packages can't handle the new version yet.

**Comment by globin** [comment]
Yes, still wip

**Comment by vcunat** [comment]
Perhaps the package updates are worth merging regardless of default openssl version?

**Comment by globin** [comment]
Was planning to finish this in the next few days anyway so not sure if it's worth it.

**Comment by globin** [comment]
Retargeting for after 17.03 as too much stuff is breaking and not easily fixable due to openssl being propagated sometimes and stuff (PHP) trying to link to both versions due to dependencies.

I'll merge the version bumps to staging.

**Comment by vcunat** [comment]
BTW, #24574 claims to fix ntp
> * [Bug 3363] Support for openssl-1.1.0 without compatibility modes
>   - rework of patch set from <ntp.org@eroen.eu>. <perlinger@ntp.org>

**Comment by vcunat** [comment]
Still far from finishing this monster? Perhaps some stuff would easy to get to staging or master directly? (like those various updates)

**Comment by FRidh** [comment]
Considering the amount of failures is already quite low (basically at the noise level) it would seem to me this is ready to go in staging. What do you think @globin?

**Comment by globin** [comment]
I'd prefer fixing up a but more, i.e. everything depending on qt 5.6 is broken..

**Comment by FRidh** [comment]
@globin Qt 5.6 seems to be broken on staging as well.
https://nix-cache.s3.amazonaws.com/log/8hya81bnhab2b0mps9amld1ma989hz44-qtbase-5.6.2.drv

**Comment by globin** [comment]
Mostly reminder to myself: erlang < R20 needs patching

**Comment by vcunat** [comment]
I wonder, do you still plan to have this in 17.09?

**Comment by fpletz** [comment]
Bumped the milestone to 18.03. We're not expecting this to be ready for 17.09.

**Comment by copumpkin** [comment]
`xar` should be happy again. I even checked the box! 😄

**Comment by NeQuissimus** [comment]
I wonder how much we should really care about old Erlang and Ruby versions...

**Comment by peti** [comment]
I'm currently updating this branch to fit on top of the current `master` @ b3970e7f5c2368fa9063aacbc2d6c99c62f78e9c. I'll push a new version in a couple of hours and report my findings.

**Comment by peti** [comment]
I force-pushed a freshly rebased version of this branch (b3197666c83ab10e04f24e4cc0d3d87897e6787b) on top of the current `master`. The original branch contents is still available for reference at https://github.com/peti/nixpkgs/tree/openssl-1.1-pre-rebase (5791eff8d80c786d26837fc64980b19330001e22). I'll wait for Hydra to deliver a preliminary verdict on the current state of affairs.

**Comment by vcunat** [comment]
Bigger breakages I can see ATM (some might only appear with updated master):
- [ ] subversion
- [ ] openssh &ndash; update conflicting with the huge openssl-1.1 patch

**Comment by vcunat** [comment]
Let me take the baton for this evening, then I'll make Hydra re-evaluate with my fixes, and we can hopefully merge soon-ish.

**Comment by vcunat** [comment]
I now see `qt5*` as significant problems to solve.  They're probably not very good cases to override to different openssl version, due to being a frequent dependency. (I expect collisions if two versions got into one process.)

**Comment by vcunat** [comment]
I didn't notice that the default 5.10 builds out of the box.  I patched 5.9 in d9e691822, and I expect 5.6 can be removed?  What maintainers think of the removal? @ttuegel, @qknight, @periklis.

**Comment by vcunat** [comment]
We might be approaching a point where we could merge and leave the rest to the community &ndash; roughly 150 [build regressions on x86_64-linux](https://hydra.nixos.org/eval/1433051?filter=x86_64-linux&compare=1432644&full=#tabs-now-fail) (almost nothing tried on darwin yet).

**Comment by peti** [comment]
> We might be approaching a point where we could merge and leave the rest to the community.

That's what I thought, too. The current effort is quite comprehensive already, and expecting the handful of contributors to this branch to fix *all of Nixpkgs* doesn't seem realistic.

**Comment by vcunat** [comment]
@globin: do you remember how the **ruby** breakage manifests?  All the versions seem to build and respond to `--help` and `--version`, and I can't see any comment fixing it (except for switch to 2.4 by default).  Hidden breakages tend to be unpleasant.

**Comment by vcunat** [comment]
The same goes for **erlang** (<R20?)

**Review by peti — APPROVED** [review]
I tested several builds locally and everything worked fine.

👍 for merging.

**Comment by periklis** [comment]
@vcunat Can't we throw this PR against a darwin slave to check out the regressions?

**Comment by peti** [comment]
@periklis, the Hydra job *does* build Darwin.

**Comment by periklis** [comment]
Ok that's what i thought. So then @vcunat means, "nothing tried on Darwin" as "nothing tried on darwin on runtime", right?

**Comment by vcunat** [comment]
No, Hydra's been rather overloaded on Darwin side, so it didn't get to building many of the builds.

**Comment by vcunat** [comment]
ruby: @zimbatm, erlang: @gleber (from `.github/CODEOWNERS`).
tl;dr: older ruby/erlang versions possibly have problems with openssl-1.1.0.  What do you think?

**Comment by NeQuissimus** [comment]
Why not just get rid of Ruby 2.3? 2.4 and 2.5 have been out for a bit...

**Comment by vcunat** [comment]
@NeQuissimus: see https://github.com/NixOS/nixpkgs/pull/34927

**Comment by globin** [comment]
Please do not merge this right now, I want to go through this again first and maybe clean up stuff.

**Comment by globin** [comment]
Wrt to ruby at least, the older versions don't build the ssl extension as it "cannot find openssl" which doesn't break the build but the world of ruby libraries trying to import ssl.

**Comment by zimbatm** [comment]
this PR is now conflicting, probably because of the ruby changes introduced in #34927

**Comment by vcunat** [comment]
Ruby changes in; Hydra's working on updated binaries, so we can test more easily.  @zimbatm: ruby 2.3 is kept, so the preferred way is to make it use older openssl?  I suppose "we" don't like the current state in this branch: succeeding build with no openssl support.

**Comment by zimbatm** [comment]
yes, are we keeping a older openssl around? setting the version for ruby 2.3 would be great.

**Comment by vcunat** [comment]
Of course we are.  A dozen packages keep pinned to `1.0.2`.

**Comment by globin** [comment]
We have to be very(!) careful with pinning stuff to 1.0.2!! If something is linked to both versions it may fail at runtime, if possible nearly everything should be 1.1. That is one reason why I'm so reluctant in rushing this. This should probably not go into 18.03

**Comment by vcunat** [comment]
For c-like stuff, if you have both in link-closure, the build will fail during linking, as there are symbols resolvable to both versions.

**Comment by peti** [comment]
> This should probably not go into 18.03.

Actually, this PR should be merged as quickly as possible so that any remaining issues can be discovered and fixed utilizing more eyes and hands that just the 2-3 people who are working on this branch.

**Comment by vcunat** [comment]
I'm certain that **if** it should get to 18.03, it needs to get to master soon (several days at most).

A similar thing holds for gcc-7, as we branch 18.03 in about two weeks, though there I'm confident much more is catchable during build-time.  With openssl the "dynamic" languages might give more trouble.

**Comment by gleber** [comment]
Erlang <R20 is failing due to a configure error:
```
checking for OpenSSL kerberos 5 support... yes
checking for krb5.h in standard locations... not found
configure: WARNING: OpenSSL is configured for kerberos but no krb5.h found
```
Which disables the Erlang stdlib `crypto` library, which almost everyone uses:
```
crypto         : OpenSSL is configured for kerberos but no krb5.h found
```

**Comment by gleber** [comment]
The best summary I've found about Erlang and OpenSSL 1.1 is here:
https://www.bountysource.com/issues/48736281-erlang-doesn-t-build-correctly-for-certain-versions-of-openssl-1-1-x

TL;DR: Erlang <R20 does not and probably will not support OpenSSL 1.1. So either we drop older Erlangs or we pin them to previous OpenSSL version.

**Comment by peti** [comment]
So ... what do we do about this PR now? Plenty of merge conflicts have popped up over the last 2 months of inactivity and I wonder how to move this effort forward.

**Comment by copumpkin** [comment]
It seems like a lot of the changes could be comfortably put into master without bumping the default OpenSSL version. Like adding `pkgconfig` to svn, making packages that rely strongly on OpenSSL 1.0.2 specify explicitly that in their dependencies, etc.

So perhaps easiest to split it up and put the "no regrets" changes directly into master, and then what's left is the core of changes that rely on OpenSSL 1.1 being the default?

**Comment by peti** [comment]
@matthewbauer, huh? Why did you do that?

**Comment by matthewbauer** [comment]
Sorry unintentional git issue.

**Comment by Mic92** [comment]
Archlinux did this ages ago. I am confident we can do it for 19.03!

**Comment by andir** [comment]
I went through all the commits and rebased (or removed) them for the current `staging` state.

My current version can be found at https://github.com/andir/nixpkgs/tree/openssl-1.1

IIRC there used to be a jobset on hydra. Did I miss it?

**Comment by peti** [comment]
> IIRC there used to be a jobset on hydra. Did I miss it?

See https://github.com/NixOS/nixpkgs/pull/22357#issue-104256005. :smile: It's at https://hydra.nixos.org/jobset/nixos/openssl-1.1.

**Comment by peti** [comment]
@andir, I re-set this PR based on your new version. This means https://hydra.nixos.org/jobset/nixos/openssl-1.1 will be building it now.

**Comment by andir** [comment]
@peti thanks ;)

**Comment by andir** [comment]
The eval has been pending for about two hours. That is probably the case because that job is disabled? (And probably why I didn't find it in the first place, I guess).

2019-03-04-135137_913x318_scrot

@peti are you able to (re-)enable jobs?

**Comment by vcunat** [comment]
"pending" means that it's queued for evaluation.  Some of our jobsets are rather expensive to evaluate (dozens of minutes).  Now it's building already.

**Comment by peti** [comment]
I re-enabled the Hydra job (and it's completed an evaluation already, too).

**Comment by andir** [comment]
Rebased on current staging & add an updated version of the qt59 patch

**Comment by andir** [comment]
Rebased on current staging, added a few more patches and fixed the qt48 commit that I broke before.

**Comment by andir** [comment]
Rebased one more time. Was able to remove a few of the version pinnings. 

I started another hydra eval to see if things have improved. Last time I worked on this it was pretty close to be mergeable IMHO.

**Comment by globin** [comment]
Moved the jobset to the nixpkgs project in order to be able to compare it to staging:

https://hydra.nixos.org/jobset/nixpkgs/openssl-1.1

File: pkgs/top-level/perl-packages.nix
Comment (JohnAZoidberg) [inline pkgs/top-level/perl-packages.nix]: I'd recommend `substituteInPlace`. That warns when upstream changes the code and the pattern no longer matches.

Comment (globin) [reply pkgs/top-level/perl-packages.nix]: I reckon this package will never be updated again, I don't think it matters.

File: nixos/modules/services/network-filesystems/kbfs.nix
Comment (JohnAZoidberg) [inline nixos/modules/services/network-filesystems/kbfs.nix]: Can you please explain why?

Comment (lheckemann) [reply nixos/modules/services/network-filesystems/kbfs.nix]: I think this only in here because of a funky rebase that brings in lot of unrelated commits.

Comment (lheckemann) [reply nixos/modules/services/network-filesystems/kbfs.nix]: That is, this was rebased on master, which has a bunch of commits including feed9154c0ccd8b42b46cc7ddcc65e46fb08eed4 that are not on staging yet.

**Comment by domenkozar** [comment]
Closing this one as it's completely out of date.

**Comment by lheckemann** [comment]
@domenkozar, @globin seems to have rebased and started working on this again 10 days ago, hence the commit spam, so I don't think it should be closed. But @globin is probably the best person to know for sure :)

**Comment by primeos** [comment]
I also don't think this should be closed.

AFAIK we don't have another PR of this kind and this change **is very important** if we want to avoid having an end-of-life version in 19.09:

https://www.openssl.org/source/:
> Note: The latest stable version is the 1.1.1 series. This is also our Long Term Support (LTS) version, supported until 11th September 2023. Our previous LTS version (1.0.2 series) will continue to be supported until 31st December 2019 (security fixes only during the last year of support). The 1.1.0 series is currently only receiving security fixes and will go out of support on 11th September 2019. All users of 1.0.2 and 1.1.0 are encouraged to upgrade to 1.1.1 as soon as possible. The 0.9.8, 1.0.0 and 1.0.1 versions are now out of support and should not be used.

Also: I though the target version of this PR was changed from 1.1.0 to 1.1.1 (or at least that there was a statement in some comment, but I didn't find it anymore :o), but looking at the current version it still seems to be based on 1.1.0 (but I might be wrong here, I only had a quick look with the web interface and this PR is too big for that).

**Comment by primeos** [comment]
@globin would it be possible to give us a quick status update regarding this PR (also regarding my previous comment: https://github.com/NixOS/nixpkgs/pull/22357#issuecomment-519683341)?

My current interpretation:
- If the milestone change wasn't done by accident I'm assuming that this PR won't land in 19.09 anymore -> we'll have to maintain an end-of-life version without official security updates.
- The [Discourse post](https://discourse.nixos.org/t/nixos-19-09-feature-freeze/3707) and this PR still refer to OpenSSL 1.1 but this isn't an LTS version -> would it be possible to switch to 1.1.1 instead or would that require significantly more effort?

Is this PR still realistic (it is open for ~2.5 years now)? Could we possibly divide it into smaller PRs (e.g. updating the most essential/important packages first)? Would it help if this PR has more visibility (e.g. if we pin it on GitHub, mention it on Discourse, etc. -> maybe the community could help with testing and updating packages)? Or could some other approach work better (e.g. updating the OpenSSL version and pinging all maintainers of affected packages and packages that depend on an older OpenSSL version -> distributing this PR, e.g. like it was done here: https://github.com/NixOS/nixpkgs/pull/56265)?

In any case: Thanks for all of your effort, I really appreciate it! :) Of course this applies to anyone who worked on this PR!

(I hope this comment doesn't give the wrong impression. I'm glad that we have this PR, but I'm also a bit concerned as we're behind most other Linux distributions and this PR was now delayed multiple times while we desperately need it.)

**Comment by globin** [comment]
I wasn't aware they were EOL'ing at the end of this year, I'll fix up the remaining larger issues (IIRC only Erlang) and then merge, then the rest will be fixed with ZHF. I'll also try and cherry-pick the smaller parts to other PRs to get parts in faster.

This is updating to openssl 1.1.1 anyway as openssl 1.1 changed the versioning and 1.1.1 is at least API-compatible to 1.1.0.

Also I'll base this on master as it has its own hydra jobset. (I'm also continuously coordinating with @lheckemann for 19.09)

**Comment by globin** [comment]
This is good enough to be merged to master, some minor failures in leaf packages still exist, but they can either be fixed in ZHF or marked as broken if nobody cares

**Comment by globin** [comment]
I'll merge this to master and that back to staging/staging-next at around midday GMT if no one beats me to it.

**Comment by vcunat** [comment]
It's in staging-next as well, since 2e6bf42a2.

**Comment by lheckemann** [comment]
#67375 for staging (WIP)

**Comment by lopsided98** [comment]
This causes a segfault in `rtmpsink` in `gst-plugins-bad`, because `rtmpdump` depends on 1.0 while `gst-plugins-bad` depends on 1.1. See #67842 for a fix.

**Pull request merged**
